lib/devices.go

```diff
@@ -0,0 +1,238 @@
+	}
+
+	// Mark all cachedDevices as removed initially
+	for _, device := range cachedDevices {
+		device.State = StateRemoved
+		mergedMap[deviceKey(device)] = device
+	}
+	// Reset persisting devices to normal, and add new devices
+	for _, newDevice := range newDevices {
+		key := deviceKey(newDevice)
+		if existingDevice, exists := mergedMap[key]; exists {
+			// Device exists, reset its status to normal
+			existingDevice.State = StateNormal
+			mergedMap[key] = existingDevice
+		} else {
+			// Device is new, add to mergedMap
+			newDevice.State = StateAdded
+			mergedMap[key] = newDevice
+
+		}
+	}
+	// Convert map back into slice and log changes since lastMergedMap
+	merged = make([]Device, 0, len(mergedMap))
+	for key, device := range mergedMap {
+		merged = append(merged, device)
+
+		if lastDevice, exists := lastMergedMap[key]; !exists {
+			addDeviceLog(device)
+			changed = true
+		} else if device.State != lastDevice.State {
+			addDeviceLog(device)
+			changed = true
+		}
+	}
+
+	merged = sortDeviceSlice(merged)
+
+	lastMergedMap = mergedMap
+
+	return changed, merged
+}
+
+func BuildDeviceTree(devices []Device) []TreeNode {
+	roots := []TreeNode{}
+	nodes := []TreeNode{}
+	for _, dev := range devices {
+		nodes = append(nodes, makeNode(dev))
+	}
+	// Loop through each node
+	for parent := range nodes {
+		parentLen := len(nodes[parent].Path)
+		// Find this node's children
+		for child := range nodes {
+			childLen := len(nodes[child].Path)
+			if parentLen >= childLen {
+				continue
+			}
+			if isChild(nodes[parent], nodes[child]) {
+				nodes[parent].Children = append(nodes[parent].Children, nodes[child])
+			}
+		}
+		if parentLen == 0 {
+			roots = append(roots, nodes[parent])
+		}
+	}
+	return roots
+}
+
+/*
+	 Checks parent and child relationship
+		Criteria: 	Same Bus
+					Path of Child matches path of parent with 1 extra port.
+*/
+func isChild(parent TreeNode, child TreeNode) bool {
+	if len(parent.Path) != (len(child.Path) - 1) {
+		return false
+	}
+	for i := range parent.Path {
+		if parent.Path[i] != child.Path[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func makeNode(device Device) TreeNode {
+	return TreeNode{
+		Device:   device,
+		Children: []TreeNode{},
+	}
+}
+
+// Returns sorted slice of Device
+// Sorts devices by Bus length, then Path.
+func sortDeviceSlice(devices []Device) []Device {
+	sort.Slice(devices, func(i, j int) bool {
+		if devices[i].Bus != devices[j].Bus {
+			return devices[i].Bus < devices[j].Bus
+		}
+		for pathIdx := 0; pathIdx < len(devices[i].Path) && pathIdx < len(devices[j].Path); pathIdx++ {
+			if devices[i].Path[pathIdx] != devices[j].Path[pathIdx] {
+				return devices[i].Path[pathIdx] < devices[j].Path[pathIdx]
+			}
+		}
+		return len(devices[i].Path) < len(devices[j].Path)
+	})
+	return devices
+}
+
+func addDeviceLog(device Device) {
+	newLog := Log{Time: time.Now(), Text: device.Name, State: device.State}
```

Comment:
`time.Now()` has a lot of precision to it, did you detect the change *now*, or when you got the latest list of devices?